blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,546 @@
+This needs to be a separate type because thread-safe use of condition variables
+requires acquiring a lock.
+In Julia, the lock is bundled with the condition, so `lock` can simply be called
+on the condition itself:
+
+```
+lock(cond::Threads.Condition)
+while !ready
+    wait(cond)
+end
+unlock(cond)
+```
+
+As in previous versions, the standard lock to use to protect critical sections
+is `ReentrantLock`, which is now thread-safe (it was previously only used for
+synchronizing tasks).
+`Threads.SpinLock` is also available, to be used in rare circumstances where
+(1) only threads and not tasks need to be synchronized, and (2) you expect to
+hold the lock for a short time.
+`Semaphore` and `Event` are also available, completing the standard set of
+synchronization primitives.
+
+### Thread-local state
+
+Julia code naturally tends to be purely functional (no side effects or mutation),
+or only uses local mutation, so migrating to full thread-safety will hopefully
+be easy in many cases.
+But if your code uses shared state and you'd like to make it thread-safe, there
+is some work to do.
+So far we have used two kinds of approaches to this in Julia's standard library:
+synchronization (locks), and thread- or task-local state.
+Locks work well for shared resources not accessed frequently, or for resources
+that cannot be duplicated for each thread.
+
+But for high-performance code we recommend thread-local state.
+Our `psort!` routine above can be improved in this way.
+Here is a recipe.
+First, we modify the function to accept pre-allocated buffers, using a default
+argument value to allocate space automatically when the caller doesn't provide it:
+
+```
+function psort!(v, lo::Int=1, hi::Int=length(v), temps = [similar(v,cld(length(v),2)) for i = 1:Threads.nthreads()])
+```
+
+The maximum size of temporary array our mergesort needs is half the array, using
+ceiling division (`cld`) to handle odd lengths.
+We simply need to allocate one array per thread.
+Next, we modify the recursive calls to reuse the space:
+
+```
+    half = @spawn psort!(v, lo, mid, temps)
+    psort!(v, mid+1, hi, temps)
+```
+
+Finally, use the array reserved for the current thread, instead of allocating a new one:
+
+```
+    temp = temps[Threads.threadid()]
+    copyto!(temp, 1, v, lo, m-lo+1)
+```
+
+After these minor modifications, let's check performance on our large
+machine:
+
+```
+$ for n in 1 2 4 8 16; do    JULIA_NUM_THREADS=$n ./julia psort.jl; done
+  2.723312 seconds (3.07 k allocations: 152.852 MiB, 0.14% gc time)
+  1.711112 seconds (3.28 k allocations: 229.149 MiB, 0.59% gc time)
+  0.971327 seconds (3.28 k allocations: 381.737 MiB, 1.60% gc time)
+  0.782790 seconds (3.28 k allocations: 686.913 MiB, 8.63% gc time)
+  0.722063 seconds (3.33 k allocations: 1.267 GiB, 21.43% gc time)
+```
+
+Definitely faster, but we do seem to have some work to do on the
+scalability of the runtime system.
+
+### Seeding the default random number generator
```

Comment:
Move this after `### IO`?

Reply:
I put it here since I consider this something you might need to know to update code, while the IO section is more internal details.